src/System.Security.Cryptography.Algorithms/tests/AesCcmTests.cs

```diff
@@ -0,0 +1,602 @@
+
+            using (var aesCcm = new AesCcm(key))
+            {
+                Assert.Throws<ArgumentNullException>(() => aesCcm.Encrypt(nonce, plaintext, ciphertext, (byte[])null));
+                Assert.Throws<ArgumentNullException>(() => aesCcm.Decrypt(nonce, ciphertext, (byte[])null, plaintext));
+            }
+        }
+
+        [ConditionalTheory(nameof(SupportsCcm))]
+        [MemberData(nameof(GetNistCcmTestCases))]
+        public static void AesCcmNistTests(AEADTest testCase)
+        {
+            using (var aesCcm = new AesCcm(testCase.Key))
+            {
+                byte[] ciphertext = new byte[testCase.Plaintext.Length];
+                byte[] tag = new byte[testCase.Tag.Length];
+                aesCcm.Encrypt(testCase.Nonce, testCase.Plaintext, ciphertext, tag, testCase.AssociatedData);
+                Assert.Equal(testCase.Ciphertext, ciphertext);
+                Assert.Equal(testCase.Tag, tag);
+
+                byte[] plaintext = new byte[testCase.Plaintext.Length];
+                aesCcm.Decrypt(testCase.Nonce, ciphertext, tag, plaintext, testCase.AssociatedData);
+                Assert.Equal(testCase.Plaintext, plaintext);
+            }
+        }
+
+        [ConditionalTheory(nameof(SupportsCcm))]
+        [MemberData(nameof(GetNistCcmTestCases))]
+        public static void AesCcmNistTestsTamperTag(AEADTest testCase)
+        {
+            using (var aesCcm = new AesCcm(testCase.Key))
+            {
+                byte[] ciphertext = new byte[testCase.Plaintext.Length];
+                byte[] tag = new byte[testCase.Tag.Length];
+                aesCcm.Encrypt(testCase.Nonce, testCase.Plaintext, ciphertext, tag, testCase.AssociatedData);
+                Assert.Equal(testCase.Ciphertext, ciphertext);
+                Assert.Equal(testCase.Tag, tag);
+
+                tag[0] ^= 1;
+
+                byte[] plaintext = new byte[testCase.Plaintext.Length];
+                Assert.Throws<CryptographicException>(
+                    () => aesCcm.Decrypt(testCase.Nonce, ciphertext, tag, plaintext, testCase.AssociatedData));
+            }
+        }
+
+        private static IEnumerable<object[]> GetValidNonceSizes()
+        {
+            return GetValidSizes(AesCcm.NonceByteSizes);
+        }
+
+        private static IEnumerable<object[]> GetInvalidNonceSizes()
+        {
+            return GetInvalidSizes(AesCcm.NonceByteSizes);
+        }
+
+        private static IEnumerable<object[]> GetValidTagSizes()
+        {
+            return GetValidSizes(AesCcm.TagByteSizes);
+        }
+
+        private static IEnumerable<object[]> GetInvalidTagSizes()
+        {
+            return GetInvalidSizes(AesCcm.TagByteSizes);
+        }
+
+        public static IEnumerable<object[]> GetNistCcmTestCases()
+        {
+            foreach (AEADTest test in s_nistCcmTestVectorsSelectedCases)
+            {
+                yield return new object[] { test };
+            }
+        }
+
```

Comment:
Some things that I can't tell from the tests:

* During tag failure on decryption, is the output A) unmodified, B) set to zero, C) contain (some or all of) the unverified decryption?
* Does in-place encryption work? `ccm.Encrypt(nonce, buf, buf, tag);`
* Does in-place decryption work?  `ccm.Decrypt(nonce, buf, tag, buf);`  What state is it in if the tag verify failed?

Reply:
added test cases